### PR TITLE
Replaced DataLoader typing with `Iterable`; Removed `WrappedDataLoader` and the `DataLoader` Protocol

### LINKED ISSUE #424: Remove the `WrappedDataloader` and `DataLoader` protocol; replace DataLoader types with `Iterable[Batch]`
We can remove our own DataLoader type and the WrappedDataLoader, and instead type the dataloader attribute on state with `Iterable[Batch]`. This is a prerequesite for FFCV integration, which uses custom dataloader types.

Steps:

- [ ] Anything that assumes a PyTorch dataloader should use an `if isisntance(dataloader, torch.utils.data.DataLoader)` check (e.g. algorithms that insert augmentations into vision datsets)
- [ ] Add in a `BEFORE_DATALOADER` event. This event should fire immediately before the underlying dataloader iterator is `__next__`'ed. See the `DataLoaderProfiler` for an example on how to wrap the dataloader
- [ ] Re-implement that DataloaderProfiler to use the `BEFORE_DATALOADER` event. It can measure the time between BEFORE_DATALOADER and AFTER_DATALOADER.
- [ ] Remove the `WrappedDataloader` class and `DataLoader` protocol class. On the `state`, update the dataloader type to `Iterable[Batch]`.

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
- Replaced DataLoader typing with `Iterable`. This is necessary for FFCV integration. Added `isinstance` checks throughout the codebase where a `torch.utils.data.DataLoader` is required
- Removed the `WrappedDataLoader` class and `DataLoader` protocol classes. These classes are opinionated and presume a PyTorch dataloader, not a generic iterable.
- Integrated the `DataLoaderProfiler` directly into the trainer. This was much simpler than adding in the `BEFORE_DATALOADER` event (as outlined in #424), which would only be used by the profiler (algorithms would never need to run on this event, since there isn't any data yet loaded! Algorithms already have the `BATCH_START` event.) and would not guarantee accurate timing. This change was necessary as the `DataLoaderProfiler` used the now-removed `WrappedDataLoader` classes.

Closes #424

### CODE REVIEW

**Review by bandish-shah — APPROVED** [review]
LGTM, I focused on reviewing the type hints and the re-factoring of the Dataloader profiler.  It would be good to have @dskhudia's stamp on this as well to confirm requirements are met for the FFCV dataloader.

I'm not a huge fan of implementing the Dataloader profiler directly in the trainer as longer term this means users that implementing their own trainers will need to also implement this themselves.  But in all fairness this isn't something we support so it's not an issue short-term and I imagine the profiler will go through more changes anyways so this won't be the last time we move things around.

File: composer/algorithms/augmix/augmix.py [hunk 1]
```diff
@@ -262,7 +265,7 @@ def apply(self, event: Event, state: State, logger: Logger) -> None:
                                     width=self.width,
                                     alpha=self.alpha,
                                     augmentation_set=self.augmentation_set)
-        assert state.dataloader is not None, "dataloader should be defined on fit start"
+        assert isinstance(state.dataloader, torch.utils.data.DataLoader), "dataloader type checked on match()"
```
Comment (dskhudia) [inline composer/algorithms/augmix/augmix.py:268]: These (augmix/colout/randaugment) algorithms as written won't work with ffcv dataloader as it has no dataset. Checking with isinstance is fine here.

File: composer/algorithms/seq_length_warmup/seq_length_warmup.py [hunk 2]
```diff
@@ -184,6 +185,8 @@ def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
             return
 
         assert state.dataloader is not None, "dataloader should be set on AFTER_DATALOADER"
+        if not isinstance(state.dataloader, torch.utils.data.DataLoader):
```
Comment (dskhudia) [inline composer/algorithms/seq_length_warmup/seq_length_warmup.py:188]: This algorithm can work with FFCV dataloader, however, adding this check will make it fail as ffcv.Loader is not an instance of torch.utils.data.DataLoader.

Comment (ravi-mosaicml) [reply composer/algorithms/seq_length_warmup/seq_length_warmup.py]: Good point; Replaced with a manual `getattr` check.

**Comment by ravi-mosaicml** [comment]
> LGTM, I focused on reviewing the type hints and the re-factoring of the Dataloader profiler. It would be good to have @dskhudia's stamp on this as well to confirm requirements are met for the FFCV dataloader.
> 
> I'm not a huge fan of implementing the Dataloader profiler directly in the trainer as longer term this means users that implementing their own trainers will need to also implement this themselves. But in all fairness this isn't something we support so it's not an issue short-term and I imagine the profiler will go through more changes anyways so this won't be the last time we move things around.

Makes sense. We can refactor it out later when there is a need to use it outside of our trainer.

**Pull request merged**